### PR TITLE
feat: update SubstraFL example

### PR DESCRIPTION
**Pull request by ThibaultFy**
Signed-off-by: ThibaultFy <50656860+ThibaultFy@users.noreply.github.com>

## Update de SubstraFL example on Flamby dataset

This PR updates the example with the various changes on SubstraFL 0.32.0 (test at initialisation, wrapping the metric Dockerfile...) and reorganise the example section.

Things to have in mind for the review:

- The links to the docs are broken as the release 0.23.0 does not exists yet (will be release before the end of November 2022)
- The example does not run in "docker" mode as Flamby ask the user to validate the installation of the datasets when being installed in the containers. No solution have been found yet (may be done in an other PR).

### CODE REVIEW

File: integration/Substra/README.md [hunk 1]
```diff
@@ -32,4 +32,4 @@ This example runs in local mode, simulating a **federated learning** experiment.
 
 ## Run the example
 
-To run the example, launch `jupyter notebook` and run the cells of the `flamby-integration-into-substra.ipynb`
+To run the example, launch `jupyter notebook`, copy past and run the cells of the `flamby-integration-into-substra.md` file.
```
Comment (jeandut) [inline integration/Substra/README.md:35]:
```suggestion
To run the example, launch `jupyter notebook` and copy paste and run the cells of the `flamby-integration-into-substra.md` file (you can also do it in classical Python console with your favorite IDE)..
```

File: integration/Substra/flamby-integration-into-substra.md [hunk 2]
```diff
@@ -33,150 +33,95 @@ You can run the command: `pip install -r requirements.txt` to install them.
...
-permissions = Permissions(public=False, authorized_ids=ORGS_ID)
-```
+The [**Dataset**](https://docs.substra.org/en/0.23.0/documentation/concepts.html#dataset) object itself does not contain the data. The proper asset to access them is the **datasample asset**.
 
-## Adding Metrics
+A **datasample** contains a local path to the data, and the key identifying the [**Dataset**](https://docs.substra.org/en/0.23.0/documentation/concepts.html#dataset) in order to have access to the proper `opener.py` file.
 
-A metric corresponds to an algorithm used to compute the score of prediction on a **datasample**.
-Concretely, a metric corresponds to an archive *(tar or zip file)*, automatically build from:
+To interface with [**FLamby**](https://github.com/owkin/FLamby), the utilization of the opener and datasample is a bit bypassed here. [**Substra**](https://docs.substra.org/en/0.23.0/) is a library built to be deployed on a real federated networks, where the path to the data is provided to the opener in order to load and read them in a personalized way directly on the organization.
```
Comment (jeandut) [inline integration/Substra/flamby-integration-into-substra.md:77]:
```suggestion
To interface with [**FLamby**](https://github.com/owkin/FLamby), in the simple case of a single machine with clients simulated as Python objects living in the same runtime, the utilization of the opener and associated datasample is implemented as a simple lookup table to find which center lives in which client. [**Substra**](https://docs.substra.org/en/0.23.0/) is a library built to be deployed on a real federated networks, where the path to the data is provided to the opener in order to load and read the data in a personalized way on the premises of the organization.
```

File: integration/Substra/flamby-integration-into-substra.md [hunk 3]
```diff
@@ -33,150 +33,95 @@ You can run the command: `pip install -r requirements.txt` to install them.
...
-METRICS_DOCKERFILE_FILES = [
-    assets_directory / "metric" / "metrics.py",
-    assets_directory / "metric" / "Dockerfile",
-]
-
-archive_path = metric_algo.file
-with zipfile.ZipFile(archive_path, "w") as z:
-    for filepath in METRICS_DOCKERFILE_FILES:
-        z.write(filepath, arcname=filepath.name)
+As we directly load a torch dataset from Flamby, the `folders` parameters is unused and the path usually leading to the data will point out a empty folder in this example.
```
Comment (jeandut) [inline integration/Substra/flamby-integration-into-substra.md:79]:
```suggestion
As we directly load a torch dataset from Flamby, the `folders` parameters is unused and the path usually leading to the data will point out to an empty folder in this example.
```

File: integration/Substra/flamby-integration-into-substra.md [hunk 4]
```diff
@@ -220,90 +160,118 @@ for ind, org_id in enumerate(ORGS_ID):
...
+    y_true =  next(iter(dataloader))[1]
+    y_pred = np.load(predictions_path)
 
-The index generator will be used a the batch sampler of the dataset, in order to save the state of the seen samples during the training, as Federated Algorithms have a notion of `num_updates`, which forced the batch sampler of the dataset to be stateful.
+    return float(fed_tcga_brca.metric(y_true, y_pred))
 
-`NUM_UPDATES` is the number of model update between each FL strategy aggregation.
+# The Dependency object is instantiated in order to install the right libraries in
+# the Python environment of each organization.
+metric_deps = Dependency(pypi_dependencies=["torch==1.11.0","numpy==1.23.1"],
```
Comment (jeandut) [inline integration/Substra/flamby-integration-into-substra.md:200]: explain local_dependencies in the comment, it's not clear to me what it does.

File: integration/Substra/flamby-integration-into-substra.md [hunk 5]
```diff
@@ -220,90 +160,118 @@ for ind, org_id in enumerate(ORGS_ID):
...
+```
+
+## SubstraFL algo definition
 
+A SubstraFL Algo gathers all the elements that we defined that run locally in each organization. This is the only SubstraFL object that is framework specific (here PyTorch specific).
+
+The torch dataset is passed **as a class** to the [**Torch Algorithms**](https://docs.substra.org/en/0.23.0/substrafl_doc/api/algorithms.html#torch-algorithms).
+Indeed, this torch Dataset will be instantiated within the algorithm, using the opener functions as **datasamples** parameters.
+
+Concerning the [**TorchFedAvgAlgo**](https://docs.substra.org/en/0.23.0/substrafl_doc/api/algorithms.html#torchfedavgalgo) interaction with [**FLamby**](https://github.com/owkin/FLamby), we need to overwrite the `_local_predict` function, used to compute the predictions of the model. In the default `_local_predict`provided by **Substrafl**, we assume that the `__getitem__` of the dataset take into account the `is_inference` argument to only return the inputs samples.
```
Comment (jeandut) [inline integration/Substra/flamby-integration-into-substra.md:258]:
```suggestion
Concerning the [**TorchFedAvgAlgo**](https://docs.substra.org/en/0.23.0/substrafl_doc/api/algorithms.html#torchfedavgalgo) interaction with [**FLamby**](https://github.com/owkin/FLamby), we need to overwrite the `_local_predict` function, used to compute the predictions of the model. In the default `_local_predict`provided by **Substrafl**, we assume that the `__getitem__` method of the dataset has a keword argument `is_inference` to only return X and not y.
```

**Review by jeandut — CHANGES_REQUESTED** [review]
Great PR, can you just review the comments I made on the doc ?

File: integration/Substra/flamby-integration-into-substra.md (hunk 4 above)
Comment (ThibaultFy) [reply integration/Substra/flamby-integration-into-substra.md:202]: Done :)

**Review by jeandut — APPROVED** [review]
LGTM
